instrumentation/kamon-cats-io-3/src/test/scala/kamon/instrumentation/futures/cats3/CatsIoInstrumentationSpec.scala

```diff
@@ -125,6 +128,67 @@ class CatsIoInstrumentationSpec extends AnyWordSpec with Matchers with ScalaFutu
         Await.result(result, 100.seconds)
       }
     }
+
+    "must correctly propagate context across thread boundaries using parallel dispatcher" in {
+      Dispatcher[IO].use { parallelDispatcher =>
+        IO {
+          val txIdKey = Context.key[String]("txId", "")
+
+          def getFromContext: String = Kamon.currentContext().get(txIdKey)
+
+          parallelDispatcher.unsafeRunSync {
+            (0 to 100).toList
+              .parTraverse { _ =>
+                val initialTxId = UUID.randomUUID.toString
+                val newContext = Context.of(txIdKey, initialTxId)
+                IO(Kamon.storeContext(newContext)).bracket(_ =>
+                  IO.fromFuture(IO(parallelDispatcher.unsafeToFuture(IO("response"))))
+                    .map(_ => getFromContext -> initialTxId))(IO delay _.close())
+                  .map { case (real, expected) => real shouldEqual expected }
+              }
+          }
+        }
+      }.unsafeRunSync()(IORuntime.global)
+    }
+
+    "retain state when resuming from completed deferred" in {
+      val txIdKey = Context.key[String]("txId", "")
+
+      def getFromContext: String = Kamon.currentContext().get(txIdKey)
+
+      def withDeferred(d: Deferred[IO, Unit]) = {
+        val initialTxId = UUID.randomUUID.toString
+        for {
+          initContext <- IO(Context.of(txIdKey, initialTxId))
+          scope <- IO(Kamon.storeContext(initContext))
+          _ <- d.get
+          fromCtx <- IO(getFromContext)
+          _ <- IO(scope.close())
+        } yield (fromCtx, initialTxId)
+      }
+
+      def test = for {
+        txId1 <- IO(UUID.randomUUID.toString)
+        txId2 <- IO(UUID.randomUUID.toString)
+        d <- IO.deferred[Unit]
+        c1 <- IO(Context.of(txIdKey, txId1))
+        s1 <- IO(Kamon.storeContext(c1))
+        t <- withDeferred(d).start
+        c2 <- IO(Context.of(txIdKey, txId2))
+        s2 <- IO(Kamon.storeContext(c2))
+        _ <- IO.sleep(100.millis)
+        _ <- d.complete(())
+        r <- t.joinWith(IO("" -> "nomatch"))
+        _ <- IO(s2.close())
+        _ <- IO(s1.close())
+//        _ <- IO(println(r._1 + ":"))
+//        _ <- IO(println(txId1, txId2, r._2))
+        // before: r._1 == r._2
+        // now: r._1 == txId2
+      } yield r._1 shouldEqual r._2
```

Comment:
I'm not completely certain that the new behaviour here is really so wrong... When resuming from a deferred completion, there's a sense in which the lexical scope has a claim to be 'the' context of execution, but the fiber that completes the deferred value also seems to have a justifiable claim - the desired semantics pobably differ significantly in different use cases...

Of course, consistency would be nice, and if I remove the `IO.sleep(100.millis)` in this for-comprehension, then I've found that the context introduced within the lexical scope of `withDeferred` wins and the tests pass

Reply:
I would think that the vast majority of the time we would want the lexical scope to be the context? The most common use case I know of for things like this is where the deferred is completed by e.g. a database connection pool, or a "send outgoing webhooks" executor, whereas the fiber that suspended is the one that belongs to a specific request.

Reply:
Yeah you're probably right, I don't know what I was thinking. Following exactly what's going on in the cats-effect code was twisting my melon